### PR TITLE
feat(storage): Added thread safety to HybridCSR format.

### LINKED ISSUE #94: feat(storage): Add thread safety to HybridCSR format.
## Problem
The ``HybridCSR_COO`` storage format currently lacks thread safety, making it unsuitable for concurrent access in multi-threaded applications. We need to implement robust thread safety mechanisms to allow safe concurrent read and write operations.

## Current Code Analysis
The HybridCSR_COO class implements a hybrid compressed sparse row (CSR) and coordinate list (COO) storage format with:
- CSR arrays for efficient graph traversal
- COO buffer for incremental updates
- Vertex mapping and ordering structures
- Build/update mechanisms for converting between formats

## Implementation Requirements
- Concurrent Read Access: Multiple threads should be able to read graph data simultaneously
- Thread-safe Modifications: Add/update/remove operations should be safe during concurrent access
- Performance Considerations: Minimize locking overhead, especially for read operations
- Memory Consistency: Ensure all threads see consistent state after modifications

## Suggested Approach
Consider implementing one of these thread safety strategies:
- Reader-Writer Lock: Using std::shared_mutex (C++17) for reader-writer pattern
- Fine-grained Locking: Separate locks for different components (CSR, COO, vertex structures)
- Copy-on-Write: Immutable data structures with atomic reference counting
- Lock-free Techniques: For specific operations where possible

## Tasks
- Analyze access patterns and identify critical sections
- Implement appropriate locking mechanisms
- Add thread safety to all public methods
- Ensure exception safety with proper RAII patterns
- Implement comprehensive test cases
- Performance benchmarking with concurrent access patterns

## Test Requirements
Create rigorous tests in tests/HybridCSRCOO.hpp covering:
- Concurrent read operations
- Mixed read-write scenarios
- Stress testing with multiple threads
- Edge cases and race condition detection
- Performance regression testing

## Notes
- Maintain backward compatibility with existing API
- Consider memory ordering and visibility guarantees
- Document thread safety guarantees in the class interface
- Provide examples of proper concurrent usage

We welcome contributors to propose and implement the most suitable thread safety approach for this storage format.

### PR DESCRIPTION
**Pull request by me-hem**
This PR adds thread safety to `PeakStore::HybridCSR_COO` (issue #94 ) by introducing coarse-grained synchronization with `std::shared_mutex`. Write operations (i.e.,` impl_addVertex`, `impl_addEdge`, `impl_updateEdge`, `impl_removeVertex`, and `impl_clearEdges`) now use `std::unique_lock` for exclusive access, while read operations (i.e., `impl_getEdge`, `impl_doesEdgeExist`, and `impl_hasVertex`) use `std::shared_lock` to allow concurrent reads. This ensures data integrity under concurrent workloads, prevents race conditions or memory corruption, and enables safe multi-threaded access to the storage engine.

### Changes made:
- `AdjacencyList.hpp` - Fixed override methods declaration
- `HybridCSR_COO.hpp` - Modified methods to use appropriate locking for safe concurrent access
- `HybridCSRCOO.cpp` - Added comprehensive thread safety tests including concurrent reads, mixed read-write operations, stress testing with multiple threads, and performance regression tests

### CODE REVIEW

File: src/StorageEngine/HybridCSR_COO.hpp [hunk 1]
```diff
@@ -26,13 +28,21 @@ class HybridCSR_COO : public PeakStorageInterface<VertexType, EdgeType> {
...
+    if (is_built_.load(std::memory_order_acquire))
       return;
-    is_built = true;
+
+    std::unique_lock<std::shared_mutex> lock(_mtx);
+
+    if (is_built_.load(std::memory_order_relaxed))
+      return;
+
+    is_built_.store(true, std::memory_order_release);
```
Comment (SharonIV0x86) [inline src/StorageEngine/HybridCSR_COO.hpp]: Marking ``is_built_`` as true before fully building it? I feel like it should be done after the building?

**Comment by SharonIV0x86** [comment]
@me-hem I feel like the unit tests for the Hybrid format should be in a separate file under the same folder, but this new file will only hold tests related to concurrency / parallelism. If possible kindly create a new file under the same folder ``HybrdiShardConcurrent.cpp`` or any suitable or more idiomatic name you can think of.

File: src/StorageEngine/HybridCSR_COO.hpp (hunk 1 above)
Comment (me-hem) [reply src/StorageEngine/HybridCSR_COO.hpp]: Yeah it should be marked only after building.

**Comment by me-hem** [comment]
Okay, I will modify the script and make new test file for concurrent Hybrid Storage operations.

**Comment by me-hem** [comment]
@SharonIV0x86 I have pushed the changes, now you can consider it for review.

**Comment by me-hem** [comment]
@SharonIV0x86, I have updated the branch; now you can proceed with it.

**Review by SharonIV0x86 — APPROVED** [review]
Approved.

**Comment by SharonIV0x86** [comment]
@me-hem Thank you for contributing, your PR has been merged.
